src/js/stores/MesosStateStore.js

```diff
@@ -73,20 +67,34 @@ class MesosStateStore extends GetSetBaseStore {
 
   subscribe() {
     const mesosStream = container.get(MesosStreamType);
-    const parsers = pipe(...Object.values(mesosStreamParsers));
     const getMasterRequest = request(
       { type: "GET_MASTER" },
       "/mesos/api/v1?get_master"
     ).retry(3);
 
-    this.stream = mesosStream
+    const parsers = pipe(...Object.values(mesosStreamParsers));
+    const dataStream = mesosStream
       .merge(getMasterRequest)
       .map(message => parsers(this.get("lastMesosState"), JSON.parse(message)))
       .map(MesosStateUtil.flagMarathonTasks)
       .do(state => {
         CompositeState.addState(state);
         this.setState(state);
-      })
+      });
+
+    const waitStream = getMasterRequest.zip(mesosStream.take(1));
+    const eventTriggerStream = dataStream.merge(
+      // Lots of things in DCOS UI rely on the MesosStateStore emitting
+      // MESOS_STATE_CHANGE events but since we're switching to the stream there will
+      // be no events coming. To avoid a bigger refactoring I decided to go with
+      // a fake emitter that starts emitting events once the initial state came through
+      // TODO: https://jira.mesosphere.com/browse/DCOS-18277
+      Rx.Observable.interval(Config.getRefreshRate())
+    );
+
+    waitStream
```

Comment:
(new to rxjs and want to clarify if I understand this correctly):

Isn't there going to be duplicates when the original `getMasterRequest` stream emits? The `dataStream` merged this stream's events in, and the `waitStream` will also emit events when `getMasterRequest` emits. Then these two are eventually concat together, and `onStreamData` is subscribed. So when GET_MASTER updates, I would expect two events, but this isn't happening when I test in the console. What am I not understanding correctly?

Reply:
**Update:** ahh I understand! I should have read the PR description more carefully - my understanding of the code now aligns with that! 😄

Reply:
There is a crude diagram we did, maybe it helps to understand a little better. 
![img_5840](https://user-images.githubusercontent.com/10524/35034732-6e1ba856-fb6f-11e7-9c40-95e5304dc012.JPG)

- M. Req = mesos request
- EVI = Event Stream (Observable.interval(...))
- BLA = 😃  is the waitStream.concat(eventTrigger)

Reply:
@nLight do we have a reference doc on how to draw streams? I found this on marble tests but not quite what we did (https://github.com/ReactiveX/RxJS/blob/master/doc/writing-marble-tests.md).

Reply:
@Fabs thanks for the diagram! I found this as a nice way to visualize the API: http://rxmarbles.com/#concat. Once I realized the difference between concat and merge, the code was pretty intuitive to understand.

Reply:
Nice for the reference @ahoskins. I think for web, `merge`, `switch`, `concat` are the most useful stream.

I also recommend this, to go deeper.  https://frontendmasters.com/courses/asynchronous-javascript/